homeassistant/components/braviatv/media_player.py

```diff
@@ -144,77 +144,87 @@ def __init__(self, client, name, pin, unique_id, device_info, ignored_sources):
+                connected = await self.hass.async_add_executor_job(
+                    self._braviarc.connect, self._pin, CLIENTID_PREFIX, NICKNAME
+                )
+                self._need_refresh = False
             else:
-                self._state = STATE_OFF
+                connected = self._braviarc.is_connected()
+            if not connected:
+                return
 
-        except Exception as exception_instance:  # pylint: disable=broad-except
-            _LOGGER.error(exception_instance)
-            self._state = STATE_OFF
+            self._state = STATE_ON
+            if await self._async_refresh_volume():
+                if await self._async_refresh_channels():
+                    await self._async_refresh_playing_info()
+                    return
+        self._state = STATE_OFF
 
     def _get_source(self):
         """Return the name of the source."""
         for key, value in self._content_mapping.items():
             if value == self._content_uri:
                 return key
 
-    def _reset_playing_info(self):
-        self._program_name = None
-        self._channel_name = None
-        self._program_media_type = None
-        self._channel_number = None
-        self._source = None
-        self._content_uri = None
-        self._duration = None
-        self._start_date_time = None
-
-    def _refresh_volume(self):
+    async def _async_refresh_volume(self):
         """Refresh volume information."""
-        volume_info = self._braviarc.get_volume_info()
+        volume_info = await self.hass.async_add_executor_job(
+            self._braviarc.get_volume_info
+        )
         if volume_info is not None:
             self._volume = volume_info.get("volume")
             self._min_volume = volume_info.get("minVolume")
             self._max_volume = volume_info.get("maxVolume")
             self._muted = volume_info.get("mute")
+            return True
+        return False
 
-    def _refresh_channels(self):
+    async def _async_refresh_channels(self):
+        """Refresh source and channels list."""
         if not self._source_list:
-            self._content_mapping = self._braviarc.load_source_list()
+            self._content_mapping = await self.hass.async_add_executor_job(
+                self._braviarc.load_source_list
+            )
             self._source_list = []
+            if not self._content_mapping:
+                return False
             for key in self._content_mapping:
                 if key not in self._ignored_sources:
                     self._source_list.append(key)
+        return True
+
+    async def _async_refresh_playing_info(self):
+        """Refresh Playing information."""
+        playing_info = await self.hass.async_add_executor_job(
+            self._braviarc.get_playing_info
+        )
+        if not playing_info:
+            self._channel_name = "App"
+            playing_info = {}
+        self._program_name = playing_info.get("programTitle")
```

Comment:
I think that `else` is missing here.

Reply:
That was intentional to set all the other values to ```None```; however you are correct about it being wrong.  It should be:
```python
    async def _async_refresh_playing_info(self):
        """Refresh Playing information."""
        playing_info = await self.hass.async_add_executor_job(
            self._braviarc.get_playing_info
        )
        self._program_name = playing_info.get("programTitle")
        self._channel_name = playing_info.get("title")
        self._program_media_type = playing_info.get("programMediaType")
        self._channel_number = playing_info.get("dispNum")
        self._content_uri = playing_info.get("uri")
        self._source = self._get_source()
        self._duration = playing_info.get("durationSec")
        self._start_date_time = playing_info.get("startDateTime")
        if  self._channel_number is None:
            self._channel_name = "App"
```
I'll wait to see if you agree before I update it.

Reply:
When TV gets signal from external source (HDMI, AV) `dispNum` is `None`.
I think
```python
playing_info = await self.hass.async_add_executor_job(
    self._braviarc.get_playing_info
)
self._program_name = playing_info.get("programTitle")
self._channel_name = playing_info.get("title")
self._program_media_type = playing_info.get("programMediaType")
self._channel_number = playing_info.get("dispNum")
self._content_uri = playing_info.get("uri")
self._source = self._get_source()
self._duration = playing_info.get("durationSec")
self._start_date_time = playing_info.get("startDateTime")
if not playing_info:
    self._channel_name = "App"
```
is a better solution.

Reply:
Ah, that was a typo (didn't mean channel_number) -- but regardless I agree. I'll update it.